cluster-autoscaler/proposals/buffers.md

```diff
@@ -0,0 +1,567 @@
+    job or replica set)
+*   We can implement better support for popular community owned controllers like
+    leader worker set where the assumption that all pods are homogeneous is
+    incorrect
+
+### Why can’t the user just specify the number of additional nodes in the buffers that point to set of nodes (like GKE CCC or Karpenter NodePool)?
+
+In many feature requests the users wanted to “just specify an additional number
+of nodes” to provision as a spare capacity. This feature is not included in this
+design because of the complexity of how this configuration may end up working.
+
+**Example 1**: user specified a buffer of 3 nodes. 3 nodes are created and on
+one of them the scheduler puts a pod. Should we create a new node now? In case
+of balanced scheduling it would be happening a lot making this buffer not work
+as expected growing the cluster with poorly utilized nodes.
+
+**Example 2**: user specified a buffer of 3 nodes, but the node class (like CCC
+or Karpenter NodePool) defines different possible shapes of nodes as fallbacks.
+What should the size of the buffer be? Is it 3 nodes of the highest priority
+that are currently available? Or should we somehow calculate 3x the size of the
+top priority on the lower priority nodes?
+
+Instead of that there is an option to specify full buffered capacity that will
+be split into chunks and provisioned as if additional pods were to be placed in
+the cluster.
+
+That being said, if there is a use case for such configuration that would have a
+clear meaning to the users and limited number of options needed to specify it,
+it should be relatively easy to add needed fields to the NodeClassBufferSpec.
+
+### Why a CRD in autoscaling.x-k8s.io and not a core k8s api?
+
+Buffers are not a required concept in each of the clusters and what they offer
+will depend on the autoscaling capabilities of the cluster. Adding them as a CRD
+in autoscaling space will make it clear and will allow for releasing it
+independently from the core k8s.
+
+## Features to be added in next iterations
+
+The initial iteration will cover basic use cases and will allow to simplify
+balloon pod/deployment management and allow for reduction of scheduler related
+overhead of preemption. But once we launch the initial integration there are
+potential next steps:
+
+*   To keep up with kubernetes changes we will add DRA support by adding a list
+    of resource claim templates that should be added to each chunk of the
+    buffer. This way it will be possible to define not only memory and cpu
+    capacity, but also any capacity that is possible to express with
+    ResourceClaim. This is not a part of the initial scope, because it would
+    require more implementation work within the cluster autoscaler in order to
+    support virtual pods with DRA resources.
+*   More features that further help users to manage the buffers:
+    *   Buffer schedule - option to define a buffer that is active only during
```

Comment:
If Buffer supported the scale subresource, you could hook it up to things like https://keda.sh/docs/2.11/scalers/cron/ to achieve this use case.